lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -1416,53 +1416,111 @@ static EnableOp getLastEnableOp(SeqOp parent) {
+    for (auto a = A.begin(); a != A.end(); ++a) {
+      StringRef groupName = a->getValue().groupName();
+      auto b = B.find(groupName);
+      if (b == B.end())
+        continue;
+      // This is also an element in B.
+      groupNames.push_back(groupName);
+      // Since these are being pulled out, erase them.
+      rewriter.eraseOp(a->getValue());
+      rewriter.eraseOp(b->getValue());
+    }
+    // Place the IfOp and EnableOp(s) inside a parallel region, in case this
+    // IfOp is nested in a SeqOp. This avoids unintentionally sequentializing
+    // the pulled out EnableOps.
+    ParOp parOp = rewriter.create<ParOp>(ifOp.getLoc());
+    rewriter.createBlock(&parOp.getBodyRegion());
+    Block *body = parOp.getBody();
+    ifOp->remove();
+    body->push_back(ifOp);
+
+    // Pull out the intersection between these two sets, and erase their
+    // counterparts in the Then and Else regions.
+    for (StringRef groupName : groupNames)
+      rewriter.create<EnableOp>(parOp.getLoc(), groupName);
 
-  if (lastThenEnableOp == nullptr || lastElseEnableOp == nullptr)
-    return failure();
+    return success();
+  } else if (auto r1 = dyn_cast<SeqOp>(thenControl);
+             auto r2 = dyn_cast<SeqOp>(elseControl)) {
+    ///   if %a with @G {                       if %a with @G {
+    ///     seq { ... calyx.enable @A }           seq { ... }
+    ///   else {                          ->    } else {
+    ///     seq { ... calyx.enable @A }           seq { ... }
+    ///   }                                     }
+    ///                                         calyx.enable @A
+    EnableOp lastThenEnableOp = getLastEnableOp(r1),
+             lastElseEnableOp = getLastEnableOp(r2);
+
+    if (lastThenEnableOp == nullptr || lastElseEnableOp == nullptr)
+      return failure();
+    if (lastThenEnableOp.groupName() != lastElseEnableOp.groupName())
+      return failure();
 
-  if (lastThenEnableOp.groupName() != lastElseEnableOp.groupName())
-    return failure();
+    // Place the IfOp and pulled EnableOp inside a sequential region, in case
+    // this IfOp is nested in a ParOp. This avoids unintentionally parallelizing
+    // the pulled out EnableOps.
+    SeqOp seqOp = rewriter.create<SeqOp>(ifOp.getLoc());
+    rewriter.createBlock(&seqOp.getBodyRegion());
+    Block *body = seqOp.getBody();
+    ifOp->remove();
+    body->push_back(ifOp);
+    rewriter.create<EnableOp>(seqOp.getLoc(), lastThenEnableOp.groupName());
+
+    // Erase the common EnableOp from the Then and Else regions.
+    rewriter.eraseOp(lastThenEnableOp);
+    rewriter.eraseOp(lastElseEnableOp);
+    return success();
+  }
 
-  // Erase both enable operations and add group enable operation after the
-  // shared IfOp parent.
-  rewriter.setInsertionPointAfter(ifOp);
-  rewriter.create<EnableOp>(ifOp.getLoc(), lastThenEnableOp.groupName());
-  rewriter.eraseOp(lastThenEnableOp);
-  rewriter.eraseOp(lastElseEnableOp);
-  return success();
+  return failure();
```

Comment:
Again, as above, i think it would be more clear if these two canonicalizations are split out into two rewrite patterns, and this pattern match failure statement is placed as early as possible.